Implemented the Item Mapper functionality

Raf-atmire opened the pull request:
In this PR we have implemented the functionality to map Items to Collections.
This can be achieved by using the following calls:
- /api/core/items/{itemUuid}/mappedCollections/{collectionUuid} POST call to create a new mapping
- /api/core/items/{itemUuid}/mappedCollections/{collectionUuid} DELETE call to delete a mapping 
- /api/core/items/{itemUuid}/mappedCollections to retrieve all the mappedCollections for this Item
- /api/core/collection/{collectionUuid}/mappedItems to retrieve all the Items mapped to this collection for which the owningCollection is not the given collection

This functionality includes changes to the services, DAOs and controllers to achieve the functionality as well as the creation of tests to ensure that the functionality doesn't break

Jira link: https://jira.duraspace.org/browse/DS-4097
Rest contract link: https://github.com/DSpace/Rest7Contract/pull/52

Raf-atmire committed:
Implemented the Item Mapper functionality

tdonohue commented:
@Raf-atmire & @benbosman  : Note, this PR is now "unblocked" but it requires updating per the *new* REST Contract approved in https://github.com/DSpace/Rest7Contract/pull/52

Commit:
DS-4097 name change from mappingCollections to mappedCollections + data change to uri-list

MarieVerdonck committed:
Error & Status for non-successful POST/DELETE mappedCollections requests
- POST /api/core/items/<:uuid>/mappedCollections
 > 405: item is a template item
 > 422: specified collection not found or is owningCollection of item
- DELETE /api/core/items/<:uuid>/mappedCollections/<:collection_uuid>
 > 405: item is a template item
 > 422: specified collection not found or is owningCollection of item

MarieVerdonck committed:
POST mappedCollections does not work now if given collection is owningCollection of item
> Throws exception, error status 405

MarieVerdonck committed:
Tests for new expected exceptions in MappingCollection endpoint
POST/DELETE:
 405: item is a template item
 422: specified collection not found or owningCollection of the item

KevinVdV commented:
This PR is now ready for review.

MarieVerdonck committed:
- Unit tests for findByCollectionMapping (incl Pagination) & countByCollectionMapping

MarieVerdonck committed:
restored auth system everywhere and size check for items in MappedItems call ($._embedded.mappingItems)

MarieVerdonck committed:
Non-code changes, just whitelines

MarieVerdonck committed:
Undo HALResource change, all tests run locally

benbosman reviewed (approved):
With the latest changes in place, this looks good to me
The functionality matches the REST Contract
The ITs are present and cover all the use cases which seem relevant to me
The updates to the ItemService are also tested in the ItemTest (apart from the indirect tests in the REST ITs)

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingCollectionRestController.java:
```diff
@@ -0,0 +1,164 @@
+        "{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}" +
+        "/mappedCollections")
+
+public class MappingCollectionRestController {
```
Please add JavaDocs to this new class & all public methods.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java:
```diff
@@ -0,0 +1,84 @@
+@RestController
+@RequestMapping("/api/core/collections/" +
+    "{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}/mappingItems")
+public class MappingItemRestController {
```
Please add Javadocs to this new class and all public methods

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java:
```diff
@@ -0,0 +1,84 @@
+
+@RestController
+@RequestMapping("/api/core/collections/" +
+    "{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}/mappingItems")
```
It seems odd to me that this endpoint is `/collections/[uuid]/mappingItems`, while the Item endpoint is `/items/[uuid]/mappedCollections`.  Shouldn't these be named similarly?

I'd recommend renaming this to be `/collections/[uuid]/mappedItems`.  I realize that may also require minor changes to the REST Contract, but it seems more consistent

Raf-atmire replied:
Hi @tdonohue thank you for the feedback. That would indeed make a lot more sense and I'll change this in the following commit. However I had an additional question as it may make sense to rename everything from 'mappingItems' to 'mappedItems' then if we're changing the endpoint syntax. For reference: https://github.com/DSpace/DSpace/pull/2282/files#diff-c56be5b7c98dee81bed835af5bf7cb6dR45
Currently they get embedded under the 'mappingItems' title, should we change this to 'mappedItems' as well along with refactoring class names etc? Just to make it more consistent with the proposed change as that would make everything even more consistent

tdonohue replied:
@Raf-atmire : Yes, I agree and I'd be in favor of renaming all "mappingItems" to "mappedItems" for consistency.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionResourceWrapperHalLinkFactory.java:
```diff
@@ -0,0 +1,41 @@
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Component
+public class MappingCollectionResourceWrapperHalLinkFactory
```
Please add Javadocs to this new class

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionRestHalLinkFactory.java:
```diff
@@ -0,0 +1,19 @@
+
+import org.dspace.app.rest.MappingCollectionRestController;
+
+public abstract class MappingCollectionRestHalLinkFactory<T>
```
Please add Javadocs to this new class

tdonohue commented on dspace-api/src/main/java/org/dspace/content/dao/ItemDAO.java:
```diff
@@ -63,6 +63,11 @@
+    public Iterator<Item> findArchivedByCollectionExcludingOwning(Context context, Collection collection, Integer limit,
+                                                                  Integer offset) throws SQLException;
+
+    public int countArchivedByCollectionExcludingOwning(Context context, Collection collection) throws SQLException;
```
Please add JavaDocs for these two new public methods.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemResourceWrapperHalLinkFactory.java:
```diff
@@ -0,0 +1,54 @@
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Component
+public class MappingItemResourceWrapperHalLinkFactory
```
Please add Javadocs to this new class & all public methods

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemRestHalLinkFactory.java:
```diff
@@ -0,0 +1,18 @@
+
+import org.dspace.app.rest.MappingItemRestController;
+
+public abstract class MappingItemRestHalLinkFactory<T> extends HalLinkFactory<T, MappingItemRestController> {
```
Please add Javadocs to this new class

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java:
```diff
@@ -0,0 +1,50 @@
+import org.dspace.app.rest.MappingCollectionRestController;
+import org.dspace.content.Item;
+
+public class MappingCollectionRestWrapper implements RestAddressableModel {
```
Please add Javadocs to this new class & all public methods

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java:
```diff
@@ -0,0 +1,50 @@
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import org.dspace.app.rest.MappingItemRestController;
+
+public class MappingItemRestWrapper implements RestAddressableModel {
```
Please add Javadocs to this new class

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingCollectionResourceWrapper.java:
```diff
@@ -0,0 +1,45 @@
+import org.dspace.app.rest.utils.Utils;
+import org.springframework.data.domain.Pageable;
+
+public class MappingCollectionResourceWrapper extends HALResource<MappingCollectionRestWrapper> {
```
Please add Javadocs to this new class

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingItemResourceWrapper.java:
```diff
@@ -0,0 +1,51 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MappingItemResourceWrapper extends HALResource<MappingItemRestWrapper> {
```
Please add Javadocs to this new class

tdonohue reviewed (commented):
I've given this a quick review and found that most of the code is currently missing Javadocs. We'll need to get those added in before we can merge.  A few other minor comments below inline.

Raf-atmire committed:
[DS-4097] added javadocs to the requested classes

tdonohue commented:
@Raf-atmire : this seems to have a minor conflict from recent merges today.  If you can get the conflict resolved, I'll do another review tomorrow (of the newly added javadocs) and see if we can get this merged.  Thanks!

Raf-atmire committed:
Changed endpoint mapping and fixed tests accordingly

KevinVdV commented:
@tdonohue The issues have been resolved, this PR is now ready for testing.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java:
```diff
@@ -0,0 +1,111 @@
+ */
+@RestController
+@RequestMapping("/api/core/collections/" +
+    "{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}/mappedItems")
```
@benbosman  and @Raf-atmire , I've just realized this endpoint isn't described in our REST Contract: https://github.com/DSpace/Rest7Contract/blob/master/collections.md

It's not a big deal for the purposes of approving this PR.  But, we need a separate PR for the REST Contract, please. 

For the REST Contract PR, we'll also need to ensure we make it clear the `/collections/[uuid]/mappedItems` endpoint is READ-ONLY.  Item mappings can only be modified via `/items/[uuid]/mappedCollections`.   I like that approach, we just need to document it.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java:
```diff
@@ -0,0 +1,111 @@
+     * returning only items that belong to a different collection but are mapped to the given one.
+     * These Items are then encapsulated in a MappingItemResourceWrapper and returned
+     *
+     * curl -X GET http://<dspace.restUrl>/api/core/collections/{uuid}/mappingItems
```
This comment should say `mappedItems`

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java:
```diff
@@ -0,0 +1,111 @@
+     * Example:
+     * <pre>
+     * {@code
+     *      curl -X GET http://<dspace.restUrl>/api/core/collections/8b632938-77c2-487c-81f0-e804f63e68e6/mappingItems
```
This example should say `mappedItems`

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/CollectionResource.java:
```diff
@@ -22,5 +22,6 @@
     public CollectionResource(CollectionRest collection, Utils utils, String... rels) {
         super(collection, utils, rels);
         add(utils.linkToSubResource(collection, CollectionRest.LICENSE));
+        add(utils.linkToSubResource(collection, "mappingItems"));
```
This should say "mappedItems"

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/hateoas/MappingItemResourceWrapper.java:
```diff
@@ -0,0 +1,55 @@
+            itemResources.add(new ItemResource(itemRest, utils));
+        }
+        this.itemResources = itemResources;
+        embedResource("mappingItems", itemResources);
```
This should say "mappedItems"

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingCollectionRestController.java:
```diff
@@ -0,0 +1,231 @@
+@RequestMapping("/api/core/items/" +
+        "{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}" +
+        "/mappedCollections")
+public class MappingCollectionRestController {
```
Shouldn't this class be named `MappedCollectionRestController`?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/MappingItemRestController.java:
```diff
@@ -0,0 +1,111 @@
+@RestController
+@RequestMapping("/api/core/collections/" +
+    "{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}/mappedItems")
+public class MappingItemRestController {
```
Shouldn't this class be named `MappedItemRestController`?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionResourceWrapperHalLinkFactory.java:
```diff
@@ -0,0 +1,44 @@
+ * This class' purpose is to add links to the MappingCollectionResourceWrapper objects
+ */
+@Component
+public class MappingCollectionResourceWrapperHalLinkFactory
```
Shouldn't this class be named `MappedCollectionResourceWrapperHalLinkFactory`?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingCollectionRestHalLinkFactory.java:
```diff
@@ -0,0 +1,23 @@
+ * This class acts as an abstract class for the MappingCollectionResourceWrapperHalLinkFactory to inherit from
+ * so it already has the Controller defined
+ */
+public abstract class MappingCollectionRestHalLinkFactory<T>
```
Shouldn't this class be named `MappedCollectionRestHalLinkFactory`?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemResourceWrapperHalLinkFactory.java:
```diff
@@ -0,0 +1,57 @@
+ * This class' purpose is to add links to the MappingItemResourceWrapper objects
+ */
+@Component
+public class MappingItemResourceWrapperHalLinkFactory
```
Shouldn't this class be named `MappedItemResourceWrapperHalLinkFactory`?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/link/MappingItemRestHalLinkFactory.java:
```diff
@@ -0,0 +1,22 @@
+ * This class acts as an abstract class for the MappingItemResourceWrapperHalLinkFactory to inherit from
+ * so it already has the Controller defined
+ */
+public abstract class MappingItemRestHalLinkFactory<T> extends HalLinkFactory<T, MappingItemRestController> {
```
Shouldn't this class be named `MappedItemRestHalLinkFactory`?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * The REST object that will define a list of CollectionRest objects to be returned by the REST api
+ */
+public class MappingCollectionRestWrapper implements RestAddressableModel {
```
Shouldn't this class be named `MappedCollectionRestWrapper`?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * The REST object that will define a list of ItemRest objects to be returned by the REST api
+ */
+public class MappingItemRestWrapper implements RestAddressableModel {
```
Shouldn't this class be named `MappedItemRestWrapper`?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingCollectionRestWrapper.java:
```diff
@@ -0,0 +1,53 @@
+
+    @JsonIgnore
+    private Item item;
+    public List<CollectionRest> getMappingCollectionRestList() {
```
"getMappedCollectionRestList()" / "setMappedCollectionRestList()"?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/model/MappingItemRestWrapper.java:
```diff
@@ -0,0 +1,53 @@
+        return mappingItemRestList;
+    }
+
+    public void setMappingItemRestList(List<ItemRest> mappingItemRestList) {
```
Again "Mapped" for getter/setter

tdonohue reviewed (commented):
@Raf-atmire, @KevinVdV and @benbosman : I hate to get picky here, but we still need to make this PR a bit more consistent with the REST Contract, etc.  It's a bit confusing to me that "Mapping" is used throughout class names, when the endpoints all use "Mapped".   To me, the proper terminology seems like it should be "MappedCollection" and "MappedItem" -- a "MappingCollection" or "MappingItem" isn't a clear name. 

So, I'd like to see more consistency here in the Class names.   That said, the methods in `ItemServiceImpl` can keep the "Mapping" terminology as it comes at the end... so, a "CollectionMapping" makes sense, but a "MappingCollection" is a confusing term.

Overall, honestly, I think this code is very close to merger.  It's well written code and tests.  It's really just the lack of consistent terminology that I'm having trouble with.  So, apologies if this seems really picky.  I'd just like to clean this up now, as I suspect it will be confusing to others as well.

Some terminology notes are inline. Though I may not have caught all of them.  Essentially, I'm just asking for "MappingCollection" to be renamed "MappedCollection" everywhere, and "MappingItem" to be renamed "MappedItem" everywhere.  So, hopefully it's a simple find & replace.

tdonohue commented:
@Raf-atmire : Just a note that this will need rebasing after the merger of #2459 as the `dspace-spring-rest` folder/webapp has been moved to `dspace-server-webapp`.

jpelgrims-atmire committed:
Merge branch 'master' into DS-4097_item-mapper-squashed-final

jpelgrims-atmire committed:
Rename code matching '[mM]apping[a-zA-Z]*' to '[mM]apped[a-zA-Z]*'

Rename 'mappingCollection' and similar code to 'mappedCollection', in order
to resolve Tim Donohue's feedback on the pull request.

tdonohue reviewed (approved):
👍 Looks good to me now.  Thanks @Raf-atmire and @benbosman 

Could someone create a (small) REST Contract PR to document the new `/api/core/collection/{collectionUuid}/mappedItems` Read-only endpoint created in this PR?  That endpoint is undocumented, but I agree that it's needed.  I'm fine with merging this PR, but I'd simply ask we "resync" the REST Contract in the near future.

benbosman commented:
I've created the additional PR at https://github.com/DSpace/Rest7Contract/pull/70

tdonohue commented:
Thanks @Raf-atmire and @benbosman.  Looks good now. Merging.
